### PR TITLE
sys/log: Updates to handle CBOR entries via nmgr

### PR DESCRIPTION
**Pull request by andrzej-kaczmarek**
With v3 we can now have CBOR (and binary data, but let's skip this for now...) stream as log entry and this should be properly transmitted via nmgr. This makes necessary changes to do this by sending `msg` field as valid CBOR stream from log entry, though `newtmgr` probably needs some updates to read it properly.

### CODE REVIEW

**Comment by andrzej-kaczmarek** [comment]
I also analyzed if it would be feasible to keep 'msg' as string and transfer CBOR/binary data encided in base64. The problem is that these entries can be (very) long and tinycbor does not have API that would allow to encode string provided in chunks. This means it would be necessary to encode data to some intermediate buffer and we may simply run our of memory here.

**Comment by sterlinghughes** [comment]
plus it will increase size over the air: i think binary makes more sense here.

**Comment by ccollins476ad** [comment]
Just a question: why is there a log type for CBOR?  Would the "binary" type work just as well for CBOR data?

**Comment by andrzej-kaczmarek** [comment]
@ccollins476ad it would work, but I see two advantages with separate type:
- for newtmgr transfer binary data has to be base64 (or sth) encoded which increases size over the air; CBOR can be just included as-is
- any reader can immediately display these data without extra information what's exactly inside binary log entry (can be e.g. used for pretty-printing).

**Comment by ccollins476ad** [comment]
Thanks for responding, Andrzej.

I don't believe the first point is correct; we can send binary data in a newtmgr response.  Perhaps you're thinking of the serial transport, which requires responses to be base64 encoded?

I see what you mean in the second point - it could be useful for newtmgr clients to automatically know how to display a response.  On the other hand, I would expect the consumer of logs to already know how to parse them by convention.

**Comment by andrzej-kaczmarek** [comment]
Yes, the data over newtmgr can be binary but actually I had log entries that are sent in response over newtmgr in mind - sorry for confusion. In this case, each log entry included in response has 'msg' parameter which is the log message and it is always encoded to CBOR response as string by log_nmgr.c (which was a fair assumption given that only string entries were logged). Now, I'd like this 'msg' to be binary data which I don't think is possible without extra encoding in general, but since the response itself is CBOR data, the log entry of type CBOR can be just appended there. In JSON notation (kind of) instead of
`{ 'msg' : 'text log entry }`
it would be:
`{ 'msg' : { 'foo' : 1, 'bar' : 'test' } }`
where `{ 'foo' : 1, 'bar' : 'test' }` is contents of log entry. I believe `newtmgr` does not understand such response since it seems to be stuck trying to receive it so it would need to check extra 'type' parameter and act accordingly. Nothing changes for string logs - they are handled exactly as previously.

HTH :-)

File: sys/log/full/src/log_nmgr.c [hunk 1]
```diff
@@ -99,6 +105,35 @@ log_nmgr_encode_entry(struct log *log, struct log_offset *log_offset,
...
+        dlen = len - sizeof(ueh);
+        break;
+    case LOG_ETYPE_BINARY:
+        /* XXX just some placeholder for now, perhaps should be base64 encoded? */
+        strcpy(data, "<binary>");
+        dlen = strlen(data);
+        break;
+    default:
+        assert(0);
+        strcpy(data, "<??\?>");
```
Comment (ccollins476ad) [inline sys/log/full/src/log_nmgr.c]: Accidental backslash? :)

**Comment by andrzej-kaczmarek** [comment]
@ccollins476ad Fixed. Sorry it took so long, I totally forgot about this PR...

File: sys/log/full/src/log_nmgr.c (hunk 1 above)
Comment (andrzej-kaczmarek) [reply sys/log/full/src/log_nmgr.c]: It's there to prevent `??>` being interpreted as trigraph. I could have used some other string though...

Comment (ccollins476ad) [reply sys/log/full/src/log_nmgr.c]: Oh, right.  I totally missed that.

**Pull request merged**